tests/test_fntrace.py

```diff
@@ -0,0 +1,25 @@
+import pytest
+
+from collections import defaultdict
+from typing import Dict
+
+from polytracker import taint_dag, ProgramTrace
+
+
+@pytest.mark.program_trace("test_fntrace.cpp")
+def test_fn_headers(program_trace: ProgramTrace):
+    assert isinstance(program_trace, taint_dag.TDProgramTrace)
+    functions = list(program_trace.tdfile.fn_headers)
+    names = set(map(lambda f: f[0], functions))
+    assert names == set(["main", "_Z9factoriali"])
+
+
+@pytest.mark.program_trace("test_fntrace.cpp")
+def test_fntrace(program_trace: ProgramTrace):
+    assert isinstance(program_trace, taint_dag.TDProgramTrace)
+    events = list(program_trace.tdfile.events)
+    assert len(events) == 10
+    kinds: Dict[taint_dag.TDEvent.Kind, int] = defaultdict(int)
+    for e in events:
```

Comment:
I feel like the order might not be consistent across optimization levels due to tail-call optimization. I suppose I could add another function or separate C source that doesn't recurse and check the order in that one.

Reply:
Hmm... it would be nice if we could detect and handle tail-calls. Haven't thought about exactly what it would mean for us, but if we loose the return at one level we would easily get out of synch I guess.

Reply:
Actually I'm not sure about the tail-call comment. I know in this particular case that the order was weird, but didn't look into why and just assumed it's probably tail-call optimization.